server/py/framework/utils/db/utils.py

```diff
@@ -0,0 +1,362 @@
+# limitations under the License.
+
+import importlib
+import os
+import re
+from collections.abc import Mapping, Sequence
+from typing import Any, Optional, Union
+from urllib.parse import parse_qs, urlparse
+
+import sqlalchemy
+
+import mlrun.common.db.dialects
+import mlrun.errors
+import mlrun.utils
+
+_DEFAULT_DRIVER_FOR_DIALECT: dict[str, str] = {
+    mlrun.common.db.dialects.Dialects.MYSQL: "pymysql",
+    mlrun.common.db.dialects.Dialects.POSTGRESQL: "psycopg2",
+}
+_ALLOWED_DRIVERS: set[str] = set(_DEFAULT_DRIVER_FOR_DIALECT.values())
+_driver_cache: dict[str, Any] = {}
+
+
+class ParsedDsn:
+    _IDENTIFIER_REGEX = re.compile(r"[a-zA-Z][a-zA-Z0-9_]*")
+    _PATH_REGEX = re.compile(r"[A-Za-z0-9_\-\.\/]+")
+
+    def __init__(self, dsn: str) -> None:
+        self._dsn = dsn
+        self._parsed = urlparse(dsn)
+        self.dialect, self.driver = self._split_scheme(
+            scheme=self._parsed.scheme,
+        )
+
+        # Connection components
+        if self.dialect == mlrun.common.db.dialects.Dialects.SQLITE:
+            self.username = None
+            self.password = None
+            self.host = None
+            self.port = None
+            # SQLite DSNs ignore database path
+            self.database = None
+        else:
+            self.username = self._parsed.username
+            self.password = self._parsed.password
+            self.host = self._parsed.hostname
+            self.port = self._parsed.port
+            self.database = self._parsed.path.lstrip("/") or None
+
+        # Query configurations
+        if self._parsed.query:
+            raw_qs = parse_qs(self._parsed.query)
+            self.configurations: dict[str, Union[str, list[str]]] = {
+                key: (value[0] if len(value) == 1 else value)
+                for key, value in raw_qs.items()
+            }
+        else:
+            self.configurations = {}
+
+    def is_valid(self) -> bool:
+        """
+        Validates the DSN by:
+        1. Ensuring dialect is supported
+        2. Checking driver identifier format
+        3. For SQLite: always valid here
+        4. For others: validating database path/name and other components
+        """
+        if (
+            not self.dialect
+            or self.dialect not in mlrun.common.db.dialects.Dialects.all()
+        ):
+            return False
+        if self.driver and not self._IDENTIFIER_REGEX.fullmatch(self.driver):
+            return False
+        raw_path = self._parsed.path.lstrip("/")
+        if self.dialect == mlrun.common.db.dialects.Dialects.SQLITE:
+            # SQLite DSNs: validate optional file path or :memory:
+            if not raw_path or raw_path == ":memory:":
+                return True
+        else:
+            if not self._PATH_REGEX.fullmatch(raw_path):
+                return False
+        if self.dialect == mlrun.common.db.dialects.Dialects.SQLITE:
+            return True
```

Comment:
This kinda renders the previous `if not raw_path or raw_path == ":memory:":` useless, as you always return true if `self.dialect == mlrun.common.db.dialects.Dialects.SQLITE` whether the raw path is validated or not